libs/preallocate/src/main/java/org/elasticsearch/preallocate/MacOsPreallocator.java

```diff
@@ -79,25 +95,37 @@ static class Fcntl {
             @SuppressWarnings("unused")
             private static final int F_VOLPOSMODE = 4; // allocate from the volume offset
 
-            public static final class FStore extends Structure implements Structure.ByReference {
-                public int fst_flags = 0;
-                public int fst_posmode = 0;
-                public NativeLong fst_offset = new NativeLong(0);
-                public NativeLong fst_length = new NativeLong(0);
-                @SuppressWarnings("unused")
-                public NativeLong fst_bytesalloc = new NativeLong(0);
-
-                @Override
-                protected List<String> getFieldOrder() {
-                    return Arrays.asList("fst_flags", "fst_posmode", "fst_offset", "fst_length", "fst_bytesalloc");
+            public static final class FStore {
+                final Memory memory = new Memory(32);
+
+                public void setFlags(int flags) {
+                    memory.setInt(0, flags);
+                }
+
+                public void setPosmode(int posmode) {
+                    memory.setInt(4, posmode);
+                }
+
+                public void setOffset(long offset) {
+                    memory.setLong(8, offset);
+                }
+
+                public void setLength(long length) {
+                    memory.setLong(16, length);
+                }
+
+                public void getBytesalloc() {
+                    memory.getLong(24);
                 }
 
             }
         }
 
-        static native int fcntl(int fd, int cmd, Fcntl.FStore fst);
+        private interface NativeFunctions extends Library {
+            int fcntl(int fd, int cmd, Object... args);
```

Comment:
I think this works (I glanced at the test failures, and I haven't seen them failing here), but I'm curious why this has a Object arg - and a variadic one

Reply:
The function is variadic:
https://man7.org/linux/man-pages/man2/fcntl.2.html

JNA (and JDK's FFI) note that variadic functions should be defined as variadic because the calling convention can be different on some systems. It was incorrect before, I'm fixing this here (fcntl never actually worked, there were errors on macos since this was untested).

Reply:
Never thought a syscall could have variadic args and did not check, makes totally sense now.

Reply:
The `syscall` function itself is variadic too ;)